util/types/json/functions.go

```diff
@@ -165,3 +165,117 @@ func extract(j JSON, pathExpr PathExpression) (ret []JSON) {
 	}
 	return
 }
+
+// autoWrapAsArray wraps input JSON into an array if needed.
+func autoWrapAsArray(j JSON, hintLength int) JSON {
+	jnew := CreateJSON(nil)
+	jnew.typeCode = typeCodeArray
+	jnew.array = make([]JSON, 0, hintLength)
+	jnew.array = append(jnew.array, j)
+	return jnew
+}
+
+// Merge merges suffixes into j according the following rules:
+// 1) adjacent arrays are merged to a single array;
+// 2) adjacent object are merged to a single object;
+// 3) a scalar value is autowrapped as an array before merge;
+// 4) an adjacent array and object are merged by autowrapping the object as an array.
+func (j JSON) Merge(suffixes []JSON) JSON {
+	if j.typeCode != typeCodeArray && j.typeCode != typeCodeObject {
+		j = autoWrapAsArray(j, len(suffixes)+1)
+	}
+	for i := 0; i < len(suffixes); i++ {
+		suffix := suffixes[i]
+		switch j.typeCode {
+		case typeCodeArray:
+			if suffix.typeCode == typeCodeArray {
+				// rule (1)
+				j.array = append(j.array, suffix.array...)
+			} else {
+				// rule (3), (4)
+				j.array = append(j.array, suffix)
+			}
+		case typeCodeObject:
+			if suffix.typeCode == typeCodeObject {
+				// rule (2)
+				for key := range suffix.object {
+					j.object[key] = suffix.object[key]
+				}
+			} else {
+				// rule (4), notice here we should retry that suffix again.
+				j = autoWrapAsArray(j, len(suffixes)+1-i)
+				i--
+			}
+		}
+	}
+	return j
+}
+
+// ModifyType is for modify a JSON. There are three valid values:
+// ModifyInsert, ModifyReplace and ModifySet.
+type ModifyType byte
+
+const (
+	// ModifyInsert is for insert a new element into a JSON.
+	ModifyInsert ModifyType = 0x01
+	// ModifyReplace is for replace an old elemList from a JSON.
+	ModifyReplace ModifyType = 0x02
+	// ModifySet = ModifyInsert | ModifyReplace
+	ModifySet ModifyType = 0x03
+)
+
+// Modify modifies a JSON object by insert, replace or set.
+// All path expressions cannot contain * or ** wildcard.
+// If any error occurs, the input won't be changed.
+func (j JSON) Modify(pathExprList []PathExpression, values []JSON, mt ModifyType) (retj JSON, err error) {
```

Comment:
I agree.
all of `json_set, json_replace, json_insert, json_rename, ...` should call `Modify`.
and in `Modify`, we can call other functions with implementation details.